File: tools/ControlRecord/ControlRecord_GNUradio/tools/ExtractNsamples16Bit.py
import numpy as np


################################################################################
class ExtractNsamples16Bit:
    
################################################################################
    def __init__(self):
        pass

################################################################################
    def Process(self, byte_buff, byte_buff_len, frame_starts, frame_cnt, N = 0):
    ########################################
    # Fixed values
        channels           = 2
        res                = 2 # resolution in bytes


    ########################################
    # Set variables    
        byte_buff   = byte_buff.view(np.uint8)

        byte_buff_len = min( byte_buff_len, byte_buff.size )

        nominal_len = N * channels * res

        if nominal_len > 0 :
            byte_buff_len = min( byte_buff_len, nominal_len )

        buff_len = byte_buff_len - byte_buff_len % (channels * res)
        int_buff = byte_buff[:buff_len].view( dtype=np.int16 ).newbyteorder()


        frame_starts_out = []
        frame_cnt_out = [] 
        ii = 0
        while ii < len(frame_starts) and frame_starts[ii] < buff_len :
            frame_starts_out.append( frame_starts[ii]/res )
            frame_cnt_out.append( frame_cnt[ii] ) 

            ii += 1

        return buff_len, int_buff, frame_starts_out, frame_cnt_out

File: trunk/tools/ControlRecord/ControlRecord_GNUradio/GNUradio_blocks/tools/ExtractNsamples16bit.cpp
#include "ExtractNsamples16bit.h"


ExtractNsamples16Bit::ExtractNsamples16Bit()
{
}


ExtractNsamples16Bit::~ExtractNsamples16Bit()
{
}


eNs16b_ret_t ExtractNsamples16Bit::Process(unsigned long			byte_buff_len, 
										std::queue<unsigned long>	frame_starts, 
										std::queue<unsigned long>	frame_cnt,
										unsigned long				N = 0)
{
	////////////////////////////////////////
    // Fixed values
	unsigned char	channels	= 2;
	unsigned char	res			= 2; // resolution in bytes


    ////////////////////////////////////////
    // Set variables    
    unsigned long nominal_len = N * channels * res;

	if ( nominal_len > 0 )
		byte_buff_len = byte_buff_len < nominal_len ? byte_buff_len : nominal_len;

	eNs16b_ret_t	ret_s;
	ret_s.buff_len = byte_buff_len - byte_buff_len % (channels * res);

	while (!frame_starts.empty() && frame_starts.front() < ret_s.buff_len)
	{
		ret_s.frame_starts.push( frame_starts.front() );
		ret_s.frame_cnt.push( frame_cnt.front() );

		frame_starts.pop();
		frame_cnt.pop();
	}

	return ret_s;
}